doc/design/python-api-1.md

```diff
@@ -0,0 +1,194 @@
+        * `download(target_dir: Union[str, Path], paths: Optional[List[str]] = None, show_progress=True, existing="error", get_metadata: Optional[bool] = None) -> ???`
+            * Use this to replace the `download()` function?
+            * The elements of `paths` are interpreted the same way as the argument to `get_assets_under_path()`
+            * `get_metadata` controls whether to create a `dandiset.yaml` file.  When `None`, it defaults to `True` if `paths` is `None` or empty, `False` otherwise.
+            * TO DO: Add an argument for controlling whether to create a directory under `target_dir` with the same name as the Dandiset ID?
+
+    * Methods of the `DraftDandiset` class (a subclass of `RemoteDandiset` used only for mutable draft versions):
+        * `set_metadata(metadata: DandisetMeta) -> None` — modifies instance in-place
+        * Methods for uploading an individual asset:
+            * `upload_asset(asset: LocalAsset, show_progress=True, existing="refesh", validation="require") -> RemoteAsset` ?
+            * `iter_upload_asset(asset: LocalAsset, existing="refesh", validation="require") -> Iterator[UploadProgressDict]` ?
+        * Methods for uploading a directory/collection of assets:
+            * `upload_assets(assets: Iterable[LocalAsset], show_progress=True, existing="refesh", validation="require") -> List[RemoteAsset]`
+
+    * Attributes & methods of the `Version` class:
+        * `identifier: str`
+        * `created: datetime`
+        * `modified: datetime`
+        * `name: str`
+        * `asset_count: int`
+        * `size: int`
+
+    * Attributes & methods of the `RemoteAsset` class (a subclass of `Asset`):
+        * `identifier: str`
+        * `path: str`
+        * `size: int`
+        * `created: datetime`
+        * `modified: datetime`
+        * `dandiset_id: str`
+        * `version_id: str`
+        * `get_metadata() -> RemoteAssetMeta`
+        * `get_digest(method: str) -> str` — Queries the metadata
+        * `get_raw_metadata() -> dict` — useful when the metadata is invalid
+        * `set_metadata(metadata: BareAssetMeta) -> None` — Assuming assets keep their identifier on metadata change, this modifies the instance in-place; otherwise, it should return the new, modified instance
+        * `delete() -> None`
+        * `download(filepath: str, show_progress=True, chunk_size=...) -> ???`
+        * `iter_download(filepath: str, chunk_size=...) -> Iterator[DownloadProgressDict]`
+
+    * Attributes of the `User` class:
+        * `username: str`
+        * `first_name: str`
+        * `last_name: str`
+        * `admin: bool`
+        * **Note:** We should ask the dandi-api developers to make the `GET/PUT /dandisets/{dandiset__pk}/users/` endpoints return full user structures, not just ones with only `username` fields, so that instances of this class can be properly populated
```

Comment:
filed https://github.com/dandi/dandi-api/issues/319 .  For now we can proceed without exposing `.admin` and then add it whenever API provides it